### ISSUE TITLE
Improve DNS SD to work with round-robin records

### ISSUE THREAD

**Issue by jacekn** [prometheus/prometheus#11614]
### Proposal

Currently there is no good way to scrape a http endpoint that is behind potentially changing or round-robin DNS.

Consider the following:
prometheus -> DNS A record responding with 3 IPs -> load balancer -> service exposing metrics

Challenges with above are:
* we can't use static_configs because it will break when DNS records change (reference https://github.com/prometheus/prometheus/issues/7552)
* dns_sd_config will discover 3 targets so duplicated metrics will be scraped. This will make aggregation error prone
* dns_sd_config will not work if the service requires host header to be sent

It would be useful to support above scenario somehow. I can think of two ways to do that:
1. Improve dns_sd_config to allow it to pick 1 IP per target and also send host header
2. Update static_configs to allow for periodic DNS refresh

But there may be other ways to solve this.

Note that a proxy server or extra DNS record might also work around the problem but I feel that those solutions are just workarounds rather than something we should recommend for monitoring systems.

**Comment by roidelapluie** [comment]
So .. I would love to find a solution to this but unfortunately changing the target of a scrape is something we should not do. If from scrape to scrape we scrape one endpoint then the other, we might have inconsistent metrics.

**Comment by aSemy** [comment]
Last week I finished a hacky little project that kind of solves this problem. I'll try and get it releasable. 

First I made sure that each instance behind the load balancer will label all metrics with `instance_id`, the value of which is an instant-specific value. (I've got a Spring Boot application, so in `application.yaml` I configured the application to tag the metrics with a random UUID that's determined on startup). 

Then I configured my workaround proxy to scrape the URL of the load balancer. The proxy will scrape the metrics endpoint, grouping the response by the UUID, until either all previously encountered instances have been scraped once, or all present instances are returned 5 times (which probably means one of the instances died). 

The responses from the instances will be cached, and will be refreshed in the background if the metrics are requested again.

The workaround proxy will expose a service-discovery endpoint, with all known instances, as well as labels that can be used to relabel the metrics so they make sense. I then configured Prometheus in `prometheus.yml` to discover these services, and scrape the workaround proxy, which will return the cached values.

**Comment by jacekn** [comment]
> So .. I would love to find a solution to this but unfortunately changing the target of a scrape is something we should not do.If from scrape to scrape we scrape one endpoint then the other, we might have inconsistent metrics.

Yes you may get inconsistent metrics if you change endpoint from scrape to scrape. But also in some cases you may get inconsistent metrics if you keep the scrape target. If DNS moves from underneath prometheus then it may be scraping random endpoint and pollute data.

So what I'm proposing is to let operators decide whether they wait to reevaluate DNS in a user-friendly way. This will let people scrape load balanced or DNS-load-balanced targets without introducing extra complexity to the monitoring infrastructure.
The default behavior could, and probably should, stay exactly as it is today.

**Comment by roidelapluie** [comment]
We could add a __meta_dns_answer_id with the ID of the answer presented to Prometheus. That would make it possible to only take the first answer.

**Comment by jacekn** [comment]
> We could add a __meta_dns_answer_id with the ID of the answer presented to Prometheus. That would make it possible to only take the first answer.

That could work.
Also maybe a config setting could do that? Something like this:
```
# "single" or "all"
# all: discover one target for each round-robin DNS record
# single: discover one target even if multiple answers are returned
[ mode: [<string>] | default = 'all' ]
```

For something like this to work we'd also need to document how to deal with the instance label.

For such thing to work for services behind vhosts we'd also need a way to send the host header.

**Comment by alexgreenbank** [comment]
Hello from the bug scrub.

This sounds sufficiently complex that a design doc might be the best way to move this forward. With no activity on this issue in 2 years we are closing it.

If anyone is interested in this and wants to take this on then consider writing up a proposal in the https://github.com/prometheus/proposals repo.
